File: apps/backend/src/routes/social-validation.ts
import { FastifyInstance } from 'fastify';
import { createClient } from '@supabase/supabase-js';
import crypto from 'crypto';
import { runPipeline, sveStatusStore } from '../tools/social-validation-engine/services/orchestrator';

// ─── Supabase admin client ─────────────────────────────────────────────────────
const supabaseAdmin =
  process.env.SUPABASE_URL && process.env.SUPABASE_SERVICE_ROLE_KEY
    ? createClient(process.env.SUPABASE_URL, process.env.SUPABASE_SERVICE_ROLE_KEY)
    : null;

// ─── Routes ───────────────────────────────────────────────────────────────────

export default async function socialValidationRoutes(app: FastifyInstance) {

  // ── POST /social-validation ──────────────────────────────────────────────────
  // Accepts: { ideaText, ideaName, targetAudience, contactName, contactEmail }
  // Returns: { id, status: 'pending' } — starts background SVE pipeline
  app.post('/social-validation', async (request, reply) => {
    try {
      const body = request.body as any;

      const {
        ideaText,
        ideaName,
        targetAudience,
        contactName,
        contactEmail
      } = body;

      if (!ideaText || String(ideaText).trim().length < 20) {
        return reply.status(400).send({ error: 'ideaText must be at least 20 characters.' });
      }

      const projectId = crypto.randomUUID();

      if (supabaseAdmin) {
        // 1. Create project record
        const { error: projectError } = await supabaseAdmin
          .from('projects')
          .insert([{
            id: projectId,
            idea_text: ideaText,
            idea_name: ideaName || contactName || 'Unnamed Idea',
            target_audience: targetAudience || 'General Audience',
            status: 'pending'
          }]);

        if (projectError) {
          request.log.error(projectError, '⚠️ Failed to create SVE project in DB.');
          return reply.status(500).send({ error: 'Failed to create SVE project record.' });
        }

        // 2. Create placeholder dd_reports entry (overall_score null = pending SVE)
        const { error: dbError } = await supabaseAdmin
          .from('dd_reports')
          .insert([{
            id: projectId,
            user_id: null,
            overall_score: null,
            verdict: 'Pending',
            report_data: {
              ideaText,
              ideaName,
              targetAudience,
              contactName: contactName || 'Anonymous',
              contactEmail: contactEmail || 'anonymous@crestcode.com',
              status: 'pending',
              toolType: 'social-validation'
            },
            is_mock: false
          }]);

        if (dbError) {
          request.log.error(dbError, '⚠️ Failed to create dd_reports placeholder.');
        }

        // 3. Kick off SVE background pipeline (fire-and-forget)
        runPipeline(projectId, ideaText).catch((err: any) => {
          request.log.error(err, `❌ SVE background pipeline failed for project ${projectId}`);
        });

        return reply.send({ id: projectId, status: 'pending' });
      }

      // ── Fallback: Supabase not configured — return mock immediately ────────────
      return reply.send({
        id: projectId,
        status: 'done',
        social_validation: {
          validation_score: 74,
          verdict: 'Moderate Demand',
          reasoning: 'Simulated SVE (no DB): Decent interest found in relevant online communities.',
          pain_points: [
            { pain_point: 'No structured feedback loop for indie devs', mentions: 12, severity: 4, confidence: 0.88, sources: [] },
            { pain_point: 'Bug prioritization is manual and time-consuming', mentions: 8, severity: 3, confidence: 0.75, sources: [] }
          ],
          competitors: [
            { name: 'UserVoice', website: 'https://uservoice.com', source_url: '', missing_features: ['In-game SDK', 'AI severity triage'], confidence: 0.8 }
          ],
          feature_requests: [
            { feature_name: 'Slack Integration for bug alerts', mentions: 9, priority: 'high' },
            { feature_name: 'One-click in-game survey widget', mentions: 6, priority: 'medium' }
          ]
        }
      });

    } catch (err: any) {
      request.log.error(err, '❌ POST /social-validation error');
      return reply.status(500).send({ error: err.message });
    }
  });

  // ── GET /social-validation/status ─────────────────────────────────────────────
  // Query: ?id=<projectId>
  // Returns: { status: 'pending' | 'done' | 'failed', failed_stage? }
  app.get('/social-validation/status', async (request, reply) => {
    try {
      const { id } = request.query as { id?: string };
      if (!id) {
        return reply.status(400).send({ error: 'Missing project ID.' });
      }

      if (!supabaseAdmin) {
        return reply.send({ status: 'done' }); // mock fallback
      }

      const { data, error } = await supabaseAdmin
        .from('projects')
        .select('status, failed_stage')
        .eq('id', id)
        .maybeSingle();

      if (error) {
        request.log.error(error, 'Supabase error fetching SVE project status');
        return reply.status(500).send({ error: error.message });
      }

      if (!data) {
        // Maybe it was already compiled — check dd_reports
        const { data: reportData } = await supabaseAdmin
          .from('dd_reports')
          .select('id, overall_score')
          .eq('id', id)
          .maybeSingle();

        if (reportData && reportData.overall_score !== null) {
          return reply.send({ status: 'done' });
        }
        return reply.status(404).send({ error: 'SVE project not found.' });
      }

      const inMemory = sveStatusStore.get(id);

      return reply.send({
        status: data.status,
        failed_stage: data.failed_stage,
        current_stage: inMemory?.current_stage || null
      });
    } catch (err: any) {
      request.log.error(err, '❌ GET /social-validation/status error');
      return reply.status(500).send({ error: err.message });
    }
  });

  // ── GET /social-validation ─────────────────────────────────────────────────────
  // Query: ?id=<projectId>
  // Returns: compiled SVE report with pain_points, competitors, feature_requests
  app.get('/social-validation', async (request, reply) => {
    try {
      const { id } = request.query as { id?: string };
      if (!id) {
        return reply.status(400).send({ error: 'Missing project ID.' });
      }

      if (!supabaseAdmin) {
        return reply.status(503).send({ error: 'Database not configured.' });
      }

      // 1. Check project is done
      const { data: projectRow } = await supabaseAdmin
        .from('projects')
        .select('status')
        .eq('id', id)
        .maybeSingle();

      if (!projectRow) {
        return reply.status(404).send({ error: 'SVE project not found.' });
      }

      if (projectRow.status !== 'done') {
        return reply.status(425).send({ error: `SVE project is still in status: ${projectRow.status}` });
      }

      // 2. Fetch SVE report record
      const { data: sveReport } = await supabaseAdmin
        .from('reports')
        .select('*')
        .eq('project_id', id)
        .order('created_at', { ascending: false })
        .limit(1)
        .maybeSingle();

      // 3. Fetch pain points with source URLs
      const { data: pps } = await supabaseAdmin
        .from('pain_points')
        .select('*')
        .eq('project_id', id);

      const painPointsOut = [];
      for (const pp of (pps || [])) {
        const { data: joinRows } = await supabaseAdmin
          .from('pain_point_sources')
          .select('source_id')
          .eq('pain_point_id', pp.id);

        const sourceIds = (joinRows || []).map((r: any) => r.source_id);
        let urls: string[] = [];
        if (sourceIds.length > 0) {
          const { data: srcRows } = await supabaseAdmin
            .from('sources')
            .select('url')
            .in('id', sourceIds);
          urls = (srcRows || []).map((s: any) => s.url);
        }
        painPointsOut.push({
          pain_point: pp.pain_point,
          mentions: pp.mentions,
          severity: pp.severity,
          confidence: pp.confidence,
          sources: urls
        });
      }

      // 4. Fetch competitors
      const { data: comps } = await supabaseAdmin
        .from('competitors')
        .select('*')
        .eq('project_id', id);

      const competitorsOut = (comps || []).map((c: any) => ({
        name: c.name,
        website: c.website,
        source_url: c.source_url,
        missing_features: c.missing_features || [],
        confidence: c.confidence
      }));

      // 5. Fetch feature requests
      const { data: feats } = await supabaseAdmin
        .from('features')
        .select('*')
        .eq('project_id', id)
        .order('mentions', { ascending: false });

      const featuresOut = (feats || []).map((f: any) => ({
        feature_name: f.feature_name,
        mentions: f.mentions,
        priority: f.priority
      }));

      // 6. Compile final payload
      const finalPayload = {
        id,
        created_at: new Date().toISOString(),
        social_validation: {
          validation_score: sveReport?.validation_score || 0,
          verdict: sveReport?.verdict || 'Unknown',
          reasoning: sveReport?.reasoning || 'Social validation completed.',
          pain_points: painPointsOut,
          competitors: competitorsOut,
          feature_requests: featuresOut
        }
      };

      // 7. Update dd_reports with compiled result
      await supabaseAdmin
        .from('dd_reports')
        .update({
          overall_score: sveReport?.validation_score || 0,
          verdict: sveReport?.verdict || 'Done',
          report_data: finalPayload,
          is_mock: false
        })
        .eq('id', id);

      return reply.send(finalPayload);
    } catch (err: any) {
      request.log.error(err, '❌ GET /social-validation error');
      return reply.status(500).send({ error: err.message });
    }
  });
}

File: apps/backend/app/routes/social_validation.py
import uuid
import asyncio
from datetime import datetime, timezone
from fastapi import APIRouter, HTTPException, BackgroundTasks, Query
from pydantic import BaseModel, Field
from typing import Optional

from app.services.supabase_client import supabase_admin
from app.tools.sve.services.orchestrator import run_pipeline, sve_status_store

router = APIRouter(prefix="/api/social-validation")

class SocialValidationRequest(BaseModel):
    ideaText: str = Field(..., min_length=20)
    ideaName: Optional[str] = None
    targetAudience: Optional[str] = None
    contactName: Optional[str] = None
    contactEmail: Optional[str] = None

@router.post("")
async def start_social_validation(req: SocialValidationRequest, background_tasks: BackgroundTasks):
    project_id = str(uuid.uuid4())
    idea_name = req.ideaName or req.contactName or "Unnamed Idea"
    target_audience = req.targetAudience or "General Audience"
    
    if supabase_admin:
        try:
            # 1. Create project record
            supabase_admin.table("projects").insert([{
                "id": project_id,
                "idea_text": req.ideaText,
                "idea_name": idea_name,
                "target_audience": target_audience,
                "status": "pending"
            }]).execute()
            
            # 2. Create placeholder dd_reports entry
            supabase_admin.table("dd_reports").insert([{
                "id": project_id,
                "user_id": None,
                "overall_score": None,
                "verdict": "Pending",
                "report_data": {
                    "ideaText": req.ideaText,
                    "ideaName": req.ideaName,
                    "targetAudience": req.targetAudience,
                    "contactName": req.contactName or "Anonymous",
                    "contactEmail": req.contactEmail or "anonymous@crestcode.com",
                    "status": "pending",
                    "toolType": "social-validation"
                },
                "is_mock": False
            }]).execute()
        except Exception as e:
            print(f"routes: failed to initialize SVE records: {e}")
            raise HTTPException(status_code=500, detail="Failed to initialize SVE records.")

        # 3. Kick off SVE background pipeline
        background_tasks.add_task(run_pipeline, project_id, req.ideaText)
        return {"id": project_id, "status": "pending"}

    # Mock Fallback if Supabase not configured
    return {
        "id": project_id,
        "status": "done",
        "social_validation": {
            "validation_score": 74,
            "verdict": "Moderate Demand",
            "reasoning": "Simulated SVE (no DB): Decent interest found in online communities.",
            "pain_points": [
                {"pain_point": "No structured feedback loop", "mentions": 12, "severity": 4, "confidence": 0.88, "sources": []}
            ],
            "competitors": [
                {"name": "UserVoice", "website": "https://uservoice.com", "source_url": "", "missing_features": ["In-game SDK"], "confidence": 0.8}
            ],
            "feature_requests": [
                {"feature_name": "Slack Integration", "mentions": 9, "priority": "high"}
            ]
        }
    }

@router.get("/status")
async def get_social_validation_status(id: str = Query(...)):
    if not supabase_admin:
        return {"status": "done"}
        
    try:
        res = supabase_admin.table("projects").select("status, failed_stage").eq("id", id).maybe_single().execute()
        data = res.data
        
        if not data:
            # Check dd_reports
            rep_res = supabase_admin.table("dd_reports").select("id, overall_score").eq("id", id).maybe_single().execute()
            if rep_res.data and rep_res.data.get("overall_score") is not None:
                return {"status": "done"}
            raise HTTPException(status_code=404, detail="SVE project not found.")

        in_memory = sve_status_store.get(id) or {}
        return {
            "status": data.get("status"),
            "failed_stage": data.get("failed_stage"),
            "current_stage": in_memory.get("current_stage")
        }
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.get("")
async def get_social_validation_report(id: str = Query(...)):
    if not supabase_admin:
        raise HTTPException(status_code=503, detail="Database not configured.")
        
    try:
        # 1. Check status is done
        p_res = supabase_admin.table("projects").select("status").eq("id", id).maybe_single().execute()
        p_data = p_res.data
        if not p_data:
            raise HTTPException(status_code=404, detail="SVE project not found.")
            
        status = p_data.get("status")
        if status != "done":
            raise HTTPException(status_code=425, detail=f"SVE project is still in status: {status}")

        # 2. Fetch SVE report
        r_res = supabase_admin.table("reports").select("*").eq("project_id", id).order("created_at", desc=True).limit(1).execute()
        sve_report = r_res.data[0] if r_res.data else {}

        # 3. Fetch pain points + sources
        pps_res = supabase_admin.table("pain_points").select("*").eq("project_id", id).execute()
        pps = pps_res.data or []
        
        pain_points_out = []
        for pp in pps:
            pp_id = pp.get("id")
            # Fetch join rows
            join_res = supabase_admin.table("pain_point_sources").select("source_id").eq("pain_point_id", pp_id).execute()
            source_ids = [row.get("source_id") for row in (join_res.data or [])]
            
            urls = []
            if source_ids:
                src_res = supabase_admin.table("sources").select("url").in_("id", source_ids).execute()
                urls = [s.get("url") for s in (src_res.data or [])]
                
            pain_points_out.append({
                "pain_point": pp.get("pain_point"),
                "mentions": pp.get("mentions"),
                "severity": pp.get("severity"),
                "confidence": pp.get("confidence"),
                "sources": urls
            })

        # 4. Fetch competitors
        comps_res = supabase_admin.table("competitors").select("*").eq("project_id", id).execute()
        competitors_out = []
        for c in (comps_res.data or []):
            competitors_out.append({
                "name": c.get("name"),
                "website": c.get("website"),
                "source_url": c.get("source_url"),
                "missing_features": c.get("missing_features") or [],
                "confidence": c.get("confidence")
            })

        # 5. Fetch feature requests
        feats_res = supabase_admin.table("features").select("*").eq("project_id", id).order("mentions", desc=True).execute()
        features_out = []
        for f in (feats_res.data or []):
            features_out.append({
                "feature_name": f.get("feature_name"),
                "mentions": f.get("mentions"),
                "priority": f.get("priority")
            })

        # 6. Compile payload
        final_payload = {
            "id": id,
            "created_at": datetime.now(timezone.utc).isoformat(),
            "social_validation": {
                "validation_score": sve_report.get("validation_score") or 0,
                "verdict": sve_report.get("verdict") or "Unknown",
                "reasoning": sve_report.get("reasoning") or "Social validation completed.",
                "pain_points": pain_points_out,
                "competitors": competitors_out,
                "feature_requests": features_out
            }
        }

        # 7. Update dd_reports
        supabase_admin.table("dd_reports").update({
            "overall_score": sve_report.get("validation_score") or 0,
            "verdict": sve_report.get("verdict") or "Done",
            "report_data": final_payload,
            "is_mock": False
        }).eq("id", id).execute()

        return final_payload
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
